plugin/src/main/java/org/opensearch/ml/processor/MLInferenceIngestProcessor.java

```diff
@@ -0,0 +1,368 @@
+        this.client = client;
+    }
+
+    /**
+     * overwrite in this execute method,
+     * when batch inference is available,
+     * to support async multiple predictions.
+     */
+    @Override
+    public void execute(IngestDocument ingestDocument, BiConsumer<IngestDocument, Exception> handler) {
+
+        List<Map<String, String>> processInputMap = inferenceProcessorAttributes.getInputMaps();
+        List<Map<String, String>> processOutputMap = inferenceProcessorAttributes.getOutputMaps();
+        int round = (processInputMap != null) ? processInputMap.size() : 0;
+
+        process_predictions(ingestDocument, handler, processInputMap, processOutputMap, 0, round);
+
+    }
+
+    @Override
+    public IngestDocument execute(IngestDocument ingestDocument) throws Exception {
+        return ingestDocument;
+    }
+
+    /**
+     * process predictions for one model for multiple rounds of predictions
+     * ingest documents after prediction rounds are completed,
+     * when no input mappings provided, default to add all fields to model input fields,
+     * when no output mapping provided, default to output as
+     * "inference_results" field (the same format as predict API)
+     */
+    private void process_predictions(
+        IngestDocument ingestDocument,
+        BiConsumer<IngestDocument, Exception> handler,
+        List<Map<String, String>> processInputMap,
+        List<Map<String, String>> processOutputMap,
+        int i,
+        int round
+    ) {
+        if (i >= round && i != 0) {
+            handler.accept(ingestDocument, null);
+        } else {
+
+            Map<String, String> modelParameters = new HashMap<>();
+
+            if (inferenceProcessorAttributes.getModelConfig() != null) {
+                modelParameters.putAll(inferenceProcessorAttributes.getModelConfig());
+            }
+            // when no input mapping is provided, default to read all fields from documents as model input
+            if (round == 0) {
+                Set<String> documentFields = ingestDocument.getSourceAndMetadata().keySet();
+                for (String field : documentFields) {
+                    getMappedModelInputFromDocuments(ingestDocument, modelParameters, field, field);
+                }
+
+            } else {
+                Map<String, String> inputMapping = processInputMap.get(i);
+                for (Map.Entry<String, String> entry : inputMapping.entrySet()) {
+                    String originalFieldName = entry.getKey();
+                    String ModelInputFieldName = entry.getValue();
+                    getMappedModelInputFromDocuments(ingestDocument, modelParameters, originalFieldName, ModelInputFieldName);
+                }
+            }
+
+            ActionRequest request = getRemoteModelInferenceRequest(modelParameters, inferenceProcessorAttributes.getModelId());
+
+            client.execute(MLPredictionTaskAction.INSTANCE, request, new ActionListener<>() {
```

Comment:
It turns out the recursive function `process_predictions` is a sequential predictions which is quite low performance, especially for remote model. It's different from foreachprocessor that doesn't have any heavy api calls. We can execute these predictions in parallel by using `GroupedActionListener`.
https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/action/support/GroupedActionListener.java

Reply:
Great call out!  I adapted using GroupedActionListener.